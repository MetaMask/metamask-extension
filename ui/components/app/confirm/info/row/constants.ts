export const TEST_ADDRESS = '0x5CfE73b6021E818B776b421B1c4Db2474086a7e1';

export enum RowAlertKey {
  EstimatedFee = 'estimatedFee',
  SigningInWith = 'signingInWith',
  RequestFrom = 'requestFrom',
  Resimulation = 'resimulation',
  Speed = 'speed',
}

export enum AlertActionKey {
  Buy = 'buy',
  ShowAdvancedGasFeeModal = 'showAdvancedGasModal',
  ShowGasFeeModal = 'showGasFeeModal',
}
