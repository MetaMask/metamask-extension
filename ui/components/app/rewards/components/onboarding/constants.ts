export const REWARDS_ONBOARD_OPTIN_LEGAL_LEARN_MORE_URL =
  'https://support.metamask.io/manage-crypto/metamask-rewards';

export const REWARDS_ONBOARD_TERMS_URL = 'https://go.metamask.io/rewards-terms';
