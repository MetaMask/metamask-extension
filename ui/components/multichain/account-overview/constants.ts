export const FUND_SLIDE = {
  id: 'fund',
  title: 'slideFundWalletTitle',
  description: 'slideFundWalletDescription',
  image: './images/slide-fund-icon.svg',
  href: 'https://portfolio.metamask.io/buy/build-quote',
};

///: BEGIN:ONLY_INCLUDE_IF(build-main,build-beta,build-flask)
export const BRIDGE_SLIDE = {
  id: 'bridge',
  title: 'slideBridgeTitle',
  description: 'slideBridgeDescription',
  image: './images/slide-bridge-icon.svg',
};
///: END:ONLY_INCLUDE_IF

export const CARD_SLIDE = {
  id: 'card',
  title: 'slideDebitCardTitle',
  description: 'slideDebitCardDescription',
  image: './images/slide-card-icon.svg',
  href: 'https://portfolio.metamask.io/card',
};

export const CASH_SLIDE = {
  id: 'cash',
  title: 'slideCashOutTitle',
  description: 'slideCashOutDescription',
  image: './images/slide-sell-icon.svg',
  href: 'https://portfolio.metamask.io/sell',
};

export const ZERO_BALANCE = '0x00';
