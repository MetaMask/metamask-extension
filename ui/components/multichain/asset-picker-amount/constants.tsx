export const LARGE_SYMBOL_LENGTH = 7;
