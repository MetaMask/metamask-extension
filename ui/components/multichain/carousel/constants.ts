export const MARGIN_VALUES = {
  SLIDE_BOTTOM: '40px',
  CONTAINER_SIDE: '16px',
  SLIDE_GAP: '2%',
  ZERO: '0',
};

export const WIDTH_VALUES = {
  FULL_WIDTH: '100%',
  STANDARD_SLIDE: '98%',
  FIRST_SLIDE: '94%',
};

export const BANNER_STYLES = {
  HEIGHT: '59px',
};

export const MAX_SLIDES = 5;
