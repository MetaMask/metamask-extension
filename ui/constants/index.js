export const ETH_DEFAULT_DECIMALS = 8;
