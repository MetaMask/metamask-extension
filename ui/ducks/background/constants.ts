import type {
  BackgroundStateProxy,
  MemStoreControllersComposedState,
} from '../../../shared/types/background';

export const initialBackgroundState: Omit<
  BackgroundStateProxy,
  keyof MemStoreControllersComposedState
> &
  Partial<{
    [ControllerName in keyof MemStoreControllersComposedState]: Partial<
      MemStoreControllersComposedState[ControllerName]
    >;
  }> = {
  isInitialized: false,
} as const;
