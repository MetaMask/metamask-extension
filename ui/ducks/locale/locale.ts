import { createSelector } from 'reselect';
import { Action } from 'redux'; // Import types for actions
import * as actionConstants from '../../store/actionConstants';
import { FALLBACK_LOCALE } from '../../../shared/modules/i18n';

/**
 * Type for the locale messages part of the state
 */
type LocaleMessagesState = {
  current?: { [key: string]: string }; // Messages for the current locale
  currentLocale?: string; // User's selected locale (unsafe for Intl API)
  en?: { [key: string]: string }; // English locale messages
};

/**
 * Payload for the SET_CURRENT_LOCALE action
 */
type SetCurrentLocaleAction = Action & {
  type: typeof actionConstants.SET_CURRENT_LOCALE;
  payload: {
    messages: { [key: string]: string };
    locale: string;
  };
};

/**
 * Type for actions that can be handled by reduceLocaleMessages
 */
type LocaleMessagesActions = SetCurrentLocaleAction;

/**
 * Initial state for localeMessages reducer
 */
const initialState: LocaleMessagesState = {};

/**
 * Reducer for localeMessages
 *
 * @param state - The current state
 * @param action - The action being dispatched
 * @returns The updated locale messages state
 */
export default function reduceLocaleMessages(
  // eslint-disable-next-line @typescript-eslint/default-param-last
  state: LocaleMessagesState = initialState,
  action: LocaleMessagesActions,
): LocaleMessagesState {
  switch (action.type) {
    case actionConstants.SET_CURRENT_LOCALE:
      return {
        ...state,
        current: action.payload.messages,
        currentLocale: action.payload.locale,
      };
    default:
      return state;
  }
}

/**
 * Type for the overall Redux state
 */
type AppState = {
  localeMessages: LocaleMessagesState;
};

/**
 * This selector returns a code from file://./../../../app/_locales/index.json.
 * NOT SAFE FOR INTL API USE. Use getIntlLocale instead for that.
 *
 * @param state - The overall state
 * @returns The user's selected locale
 */
export const getCurrentLocale = (state: AppState): string | undefined =>
  state.localeMessages.currentLocale;

/**
 * This selector returns a BCP 47 Language Tag for use with the Intl API.
 *
 * @returns The user's selected locale in BCP 47 format
 */
export const getIntlLocale = createSelector(
  getCurrentLocale,
  (locale): string =>
    Intl.getCanonicalLocales(
      locale ? locale.replace(/_/gu, '-') : FALLBACK_LOCALE,
    )[0],
);

/**
 * This selector returns the current locale messages.
 *
 * @param state - The overall state
 * @returns The current locale's messages
 */
export const getCurrentLocaleMessages = (
  state: AppState,
): Record<string, string> | undefined => state.localeMessages.current;

/**
 * This selector returns the English locale messages.
 *
 * @param state - The overall state
 * @returns The English locale's messages
 */
export const getEnLocaleMessages = (
  state: AppState,
): Record<string, string> | undefined => state.localeMessages.en;
