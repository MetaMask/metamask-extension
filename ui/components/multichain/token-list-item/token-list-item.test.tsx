import React from 'react';
import configureMockStore from 'redux-mock-store';

import { fireEvent, waitFor } from '@testing-library/react';
import { renderWithProvider } from '../../../../test/lib/render-helpers';
import { CHAIN_IDS } from '../../../../shared/constants/network';
import { getIntlLocale } from '../../../ducks/locale/locale';
import { mockNetworkState } from '../../../../test/stub/networks';
import { useSafeChains } from '../../../pages/settings/networks-tab/networks-form/use-safe-chains';
import { TokenListItem } from '.';

const state = {
  metamask: {
    ...mockNetworkState({ chainId: CHAIN_IDS.MAINNET }),
    useTokenDetection: false,
    currencyRates: {},
    preferences: {},
    internalAccounts: {
      accounts: {
        'cf8dace4-9439-4bd4-b3a8-88c821c8fcb3': {
          address: '0x0dcd5d886577d5081b0c52e242ef29e70be3e7bc',
          id: 'cf8dace4-9439-4bd4-b3a8-88c821c8fcb3',
          metadata: {
            name: 'Test Account',
            keyring: {
              type: 'HD Key Tree',
            },
          },
          options: {},
        },
      },
      selectedAccount: 'cf8dace4-9439-4bd4-b3a8-88c821c8fcb3',
    },
  },
};

const safeChainDetails = {
  chainId: '1',
  nativeCurrency: {
    symbol: 'ETH',
  },
};

let openTabSpy: jest.SpyInstance<void, [opts: { url: string }], unknown>;

jest.mock('../../../ducks/locale/locale', () => ({
  getIntlLocale: jest.fn(),
}));

jest.mock(
  '../../../pages/settings/networks-tab/networks-form/use-safe-chains',
  () => ({
    useSafeChains: jest.fn().mockReturnValue({
      safeChains: [safeChainDetails],
    }),
  }),
);

const mockGetIntlLocale = getIntlLocale;
const mockGetSafeChains = useSafeChains;

describe('TokenListItem', () => {
  beforeAll(() => {
    global.platform = { openTab: jest.fn(), closeCurrentWindow: jest.fn() };
    openTabSpy = jest.spyOn(global.platform, 'openTab');
    (mockGetIntlLocale as unknown as jest.Mock).mockReturnValue('en-US');
  });
  const props = {
    onClick: jest.fn(),
    tokenImage: '',
    title: '',
  };
  it('should render correctly', () => {
    const store = configureMockStore()(state);
    const { getByTestId, container } = renderWithProvider(
      <TokenListItem {...props} />,
      store,
    );
    expect(getByTestId('multichain-token-list-item')).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  it('should render with custom className', () => {
    const store = configureMockStore()(state);
    const { getByTestId } = renderWithProvider(
      <TokenListItem className="multichain-token-list-item-test" {...props} />,
      store,
    );
    expect(getByTestId('multichain-token-list-item')).toHaveClass(
      'multichain-token-list-item-test',
    );
  });

  it('should render crypto balance with warning scam', () => {
    const store = configureMockStore()(state);
    const propsToUse = {
      primary: '11.9751 ETH',
      isNativeCurrency: true,
      isOriginalTokenSymbol: false,
      tokenImage: '',
      title: '',
    };
    const { getByText } = renderWithProvider(
      <TokenListItem {...propsToUse} />,
      store,
    );
    expect(getByText('11.9751 ETH')).toBeInTheDocument();
  });

  it('should display warning scam modal', () => {
    const store = configureMockStore()(state);
    const propsToUse = {
      primary: '11.9751 ETH',
      isNativeCurrency: true,
      isOriginalTokenSymbol: false,
      showPercentage: true,
      tokenImage: '',
      title: '',
      tokenSymbol: 'SCAM_TOKEN',
    };
    const { getByTestId, getByText } = renderWithProvider(
      <TokenListItem {...propsToUse} />,
      store,
    );

    const warningScamModal = getByTestId('scam-warning');
    fireEvent.click(warningScamModal);

    expect(
      getByText(
        'The native token symbol does not match the expected symbol of the native token for the network with the associated chain ID. You have entered SCAM_TOKEN while the expected token symbol is ETH. Please verify you are connected to the correct chain.',
      ),
    ).toBeInTheDocument();
  });

  it('should display warning scam modal fallback when safechains fails to resolve correctly', () => {
    (mockGetSafeChains as unknown as jest.Mock).mockReturnValue([]);
    const store = configureMockStore()(state);
    const propsToUse = {
      primary: '11.9751 ETH',
      isNativeCurrency: true,
      isOriginalTokenSymbol: false,
      showPercentage: true,
      tokenImage: '',
      title: '',
      tokenSymbol: 'SCAM_TOKEN',
    };
    const { getByTestId, getByText } = renderWithProvider(
      <TokenListItem {...propsToUse} />,
      store,
    );

    const warningScamModal = getByTestId('scam-warning');
    fireEvent.click(warningScamModal);

    expect(
      getByText(
        'The native token symbol does not match the expected symbol of the native token for the network with the associated chain ID. You have entered SCAM_TOKEN while the expected token symbol is something else. Please verify you are connected to the correct chain.',
      ),
    ).toBeInTheDocument();
  });

  it('should render crypto balance', () => {
    const store = configureMockStore()({
      ...state,
      preferences: {},
    });
    const propsToUse = {
      primary: '11.9751 ETH',
      isNativeCurrency: true,
      isOriginalTokenSymbol: false,
      tokenImage: '',
      title: '',
    };

    const { getByText } = renderWithProvider(
      <TokenListItem {...propsToUse} />,
      store,
    );
    expect(getByText('11.9751 ETH')).toBeInTheDocument();
  });

  it('handles click action and fires onClick', () => {
    const store = configureMockStore()(state);
    const { queryByTestId } = renderWithProvider(
      <TokenListItem {...props} tokenImage="" title="" />,
      store,
    );

    const targetElem = queryByTestId('multichain-token-list-button');

    targetElem && fireEvent.click(targetElem);

    expect(props.onClick).toHaveBeenCalled();
  });

  it('handles clicking staking opens tab', async () => {
    const store = configureMockStore()(state);
    const { queryByTestId } = renderWithProvider(
      <TokenListItem isStakeable {...props} />,
      store,
    );

    const stakeButton = queryByTestId(
      `staking-entrypoint-${CHAIN_IDS.MAINNET}`,
    );

    expect(stakeButton).toBeInTheDocument();
    expect(stakeButton).not.toBeDisabled();

    stakeButton && fireEvent.click(stakeButton);
    expect(openTabSpy).toHaveBeenCalledTimes(1);

    await waitFor(() =>
      expect(openTabSpy).toHaveBeenCalledWith({
        url: expect.stringContaining('/stake?metamaskEntry=ext_stake_button'),
      }),
    );
  });
});
