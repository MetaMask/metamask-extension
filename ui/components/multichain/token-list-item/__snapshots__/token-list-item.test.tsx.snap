// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenListItem should render correctly 1`] = `
<div>
  <div
    class="mm-box multichain-token-list-item mm-box--display-flex mm-box--gap-4 mm-box--flex-direction-column"
    data-testid="multichain-token-list-item"
  >
    <a
      class="mm-box multichain-token-list-item__container-cell multichain-token-list-item__container-cell--clickable mm-box--padding-top-2 mm-box--padding-right-4 mm-box--padding-bottom-2 mm-box--padding-left-4 mm-box--display-flex mm-box--flex-direction-row"
      data-testid="multichain-token-list-button"
      href="#"
    >
      <div
        class="mm-box mm-badge-wrapper multichain-token-list-item__badge mm-box--margin-right-4 mm-box--display-inline-block"
      >
        <div
          class="mm-box mm-text mm-avatar-base mm-avatar-base--size-md mm-avatar-token mm-text--body-sm mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full"
        >
          ?
        </div>
        <div
          class="mm-box mm-badge-wrapper__badge-container mm-badge-wrapper__badge-container--circular-top-right"
        >
          <div
            class="mm-box mm-text mm-avatar-base mm-avatar-base--size-xs mm-avatar-network multichain-token-list-item__badge__avatar-network mm-text--body-xs mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-transparent box--border-style-solid box--border-width-1"
          >
            <img
              alt="Ethereum Mainnet logo"
              class="mm-avatar-network__network-image"
              src="./images/eth_logo.svg"
            />
          </div>
        </div>
      </div>
      <div
        class="mm-box multichain-token-list-item__container-cell--text-container mm-box--display-flex mm-box--flex-direction-column mm-box--width-full"
        style="flex-grow: 1; overflow: hidden;"
      >
        <div
          class="mm-box mm-box--display-flex mm-box--gap-1 mm-box--flex-direction-row mm-box--justify-content-space-between"
        >
          <div
            class="mm-box mm-box--display-inline-block mm-box--width-1/3"
          >
            <span
              class="mm-box mm-text mm-text--body-md mm-text--font-weight-medium mm-text--ellipsis mm-box--color-text-default"
            />
            <p
              class="mm-box mm-text mm-text--body-md mm-text--ellipsis mm-box--color-text-alternative"
              data-testid="multichain-token-list-item-token-name"
            />
          </div>
          <div
            class="mm-box mm-box--display-flex mm-box--flex-direction-column mm-box--align-items-flex-end mm-box--width-2/3"
          >
            <p
              class="mm-box mm-text mm-text--body-md mm-text--font-weight-medium mm-text--text-align-end mm-box--width-2/3 mm-box--color-text-default"
              data-testid="multichain-token-list-item-secondary-value"
            />
            <p
              class="mm-box mm-text mm-text--body-sm-medium mm-text--text-align-end mm-box--color-text-alternative"
              data-testid="multichain-token-list-item-value"
            >
               
            </p>
          </div>
        </div>
        <div
          class="mm-box mm-box--display-flex mm-box--gap-1 mm-box--flex-direction-row mm-box--justify-content-space-between"
        />
      </div>
    </a>
  </div>
</div>
`;
